Beam writes config with defaults and it should be runnable

Beam (https://github.com/LBNL-UCB-STI/beam/pull/2730) Beam writes config without defaults ([beam-tempate.conf](https://github.com/LBNL-UCB-STI/beam/blob/develop/src/main/resources/beam-template.conf)). We should include the defaults to the output too. Something like this:
```
private def writeFullConfigs(config: TypesafeConfig, outputDirectory: String): Unit = {
  val fullConfig = config.withFallback(ConfigFactory.parseResources("beam-template.conf").resolve())
  val configConciseWithoutJson = fullConfig.root().render(ConfigRenderOptions.concise().setFormatted(true).setJson(false))
  writeStringToFile(configConciseWithoutJson, new File(outputDirectory, "fullBeamConfig.conf"))
  writeStringToFile(fullConfig.root().render(), new File(outputDirectory, "fullBeamConfigJson.conf"))
}
```
allows to write both provided config and `beam-template.conf`, but then the result file is not parsable by `BeamConfig` and it fails with the errors like:
```
    Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.CommandLineWrapper.main(CommandLineWrapper.java:64)
Caused by: com.typesafe.config.ConfigException$WrongType: C:\repos\beam\output\beamville\beamville__2020-06-10_00-56-07_yec\fullBeamConfig.conf: 89: Level3 has type STRING rather than NUMBER
	at com.typesafe.config.impl.SimpleConfig.findKeyOrNull(SimpleConfig.java:163)
	at com.typesafe.config.impl.SimpleConfig.findOrNull(SimpleConfig.java:174)
	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:188)
	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:193)
	at com.typesafe.config.impl.SimpleConfig.getConfigNumber(SimpleConfig.java:223)
	at com.typesafe.config.impl.SimpleConfig.getNumber(SimpleConfig.java:229)
	at com.typesafe.config.impl.SimpleConfig.getDouble(SimpleConfig.java:245)
	at beam.sim.config.BeamConfig$Beam$Agentsim$Agents$ModalBehaviors$HighTimeSensitivity$HighCongestion$HighwayFactor$.apply(BeamConfig.scala:131)
	at beam.sim.config.BeamConfig$Beam$Agentsim$Agents$ModalBehaviors$HighTimeSensitivity$HighCongestion$.apply(BeamConfig.scala:167)
	at beam.sim.config.BeamConfig$Beam$Agentsim$Agents$ModalBehaviors$HighTimeSensitivity$.apply(BeamConfig.scala:250)
	at beam.sim.config.BeamConfig$Beam$Agentsim$Agents$ModalBehaviors$.apply(BeamConfig.scala:546)
	at beam.sim.config.BeamConfig$Beam$Agentsim$Agents$.apply(BeamConfig.scala:1389)
	at beam.sim.config.BeamConfig$Beam$Agentsim$.apply(BeamConfig.scala:1579)
	at beam.sim.config.BeamConfig$Beam$.apply(BeamConfig.scala:3206)
	at beam.sim.config.BeamConfig$.apply(BeamConfig.scala:3816)
	at beam.sim.BeamHelper.setupBeamWithConfig(BeamHelper.scala:635)
	at beam.sim.BeamHelper.setupBeamWithConfig$(BeamHelper.scala:632)
```

Find out the solution for this.

--- a/src/main/resources/beam-template.conf
+++ b/src/main/resources/beam-template.conf
@@ -105,7 +105,7 @@ beam.agentsim.agents.parking.maxSearchRadius = "double | 8046.72" // something l
 
 #TAZ params
 beam.agentsim.taz.filePath = ${beam.inputDirectory}"/taz-centers.csv"
-beam.agentsim.taz.parkingFilePath = ${beam.inputDirectory}"/taz-parking.csv"
+beam.agentsim.taz.parkingFilePath = ""
 beam.agentsim.taz.parkingStallCountScalingFactor = "double | 1.0"
 beam.agentsim.taz.parkingCostScalingFactor = "double | 1.0"
 # options: DEFAULT, PARALLEL
@@ -118,7 +118,7 @@ beam.agentsim.toll.filePath = ${beam.inputDirectory}"/toll-prices.csv"
 beam.agentsim.agents.rideHailTransit.modesToConsider = "MASS"
 # Ride Hailing Params
 beam.agentsim.agents.rideHail.initialization.initType = "PROCEDURAL" # Other possible values - FILE
-beam.agentsim.agents.rideHail.initialization.filePath = ${beam.inputDirectory}"/ride-hail-fleet.csv"
+beam.agentsim.agents.rideHail.initialization.filePath = ""
 beam.agentsim.agents.rideHail.initialization.parking.filePath = ""
 beam.agentsim.agents.rideHail.defaultBaseCost = 1.8
 beam.agentsim.agents.rideHail.defaultCostPerMile = 0.91
@@ -719,7 +719,7 @@ beam.calibration.counts {
   countsScaleFactor = 10
   writeCountsInterval = 1
   averageCountsOverIterations = 1
-  inputCountsFile = ${beam.inputDirectory}"/counts.xml"
+  inputCountsFile = ""
 }
 
 ##################################################################

--- a/src/main/resources/beam-template.conf
+++ b/src/main/resources/beam-template.conf
@@ -119,7 +119,7 @@ beam.agentsim.agents.rideHailTransit.modesToConsider = "MASS"
 # Ride Hailing Params
 beam.agentsim.agents.rideHail.initialization.initType = "PROCEDURAL" # Other possible values - FILE
 beam.agentsim.agents.rideHail.initialization.filePath = ${beam.inputDirectory}"/ride-hail-fleet.csv"
-beam.agentsim.agents.rideHail.initialization.parking.filePath = ${beam.inputDirectory}"/ride-hail-parking.csv"
+beam.agentsim.agents.rideHail.initialization.parking.filePath = ""
 beam.agentsim.agents.rideHail.defaultBaseCost = 1.8
 beam.agentsim.agents.rideHail.defaultCostPerMile = 0.91
 beam.agentsim.agents.rideHail.defaultCostPerMinute = 0.28

--- a/src/main/scala/beam/sim/BeamHelper.scala
+++ b/src/main/scala/beam/sim/BeamHelper.scala
@@ -690,7 +690,7 @@ trait BeamHelper extends LazyLogging {
       .entrySet()
       .asScala
       .collect {
-        case entry if addKey(entry.getValue.unwrapped) =>
+        case entry if shouldAddKey(entry.getValue.unwrapped) =>
           val unwrapped = entry.getValue.unwrapped()
           val paramValue = unwrapped.toString
           if (paramValue.contains("|")) {
@@ -728,7 +728,7 @@ trait BeamHelper extends LazyLogging {
     value
   }
 
-  private def addKey(value: AnyRef): Boolean = {
+  private def shouldAddKey(value: AnyRef): Boolean = {
     if ("int?" == value.toString || "double?" == value.toString || "[double]" == value.toString) {
       return false
     }

--- a/src/main/scala/beam/sim/BeamHelper.scala
+++ b/src/main/scala/beam/sim/BeamHelper.scala
@@ -714,6 +714,12 @@ trait BeamHelper extends LazyLogging {
       return value.toInt
     }
     if (paramValue.contains("double")) {
+      if ("Double.PositiveInfinity" == value) {
+        return Double.PositiveInfinity
+      }
+      if ("Double.NegativeInfinity" == value) {
+        return Double.NegativeInfinity
+      }
       return value.toDouble
     }
     if (paramValue.contains("boolean")) {

--- a/src/main/scala/beam/sim/BeamHelper.scala
+++ b/src/main/scala/beam/sim/BeamHelper.scala
@@ -39,13 +39,20 @@ import com.conveyal.r5.transit.TransportNetwork
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.module.scala.DefaultScalaModule
 import com.google.inject
-import com.typesafe.config.{ConfigFactory, ConfigRenderOptions, Config => TypesafeConfig}
+import com.typesafe.config.{
+  ConfigFactory,
+  ConfigRenderOptions,
+  ConfigResolveOptions,
+  ConfigValueFactory,
+  ConfigValueType,
+  Config => TypesafeConfig
+}
 import com.typesafe.scalalogging.LazyLogging
 import kamon.Kamon
 import org.matsim.api.core.v01.{Id, Scenario}
 import org.matsim.core.api.experimental.events.EventsManager
 import org.matsim.core.config.groups.TravelTimeCalculatorConfigGroup
-import org.matsim.core.config.{Config => MatsimConfig}
+import org.matsim.core.config.{ConfigWriter, Config => MatsimConfig}
 import org.matsim.core.controler._
 import org.matsim.core.controler.corelisteners.{ControlerDefaultCoreListenersModule, EventsHandling, PlansDumping}
 import org.matsim.core.scenario.{MutableScenario, ScenarioBuilder, ScenarioByInstanceModule, ScenarioUtils}
@@ -667,10 +674,59 @@ trait BeamHelper extends LazyLogging {
     * @param outputDirectory output folder where full configs will be generated
     */
   private def writeFullConfigs(config: TypesafeConfig, outputDirectory: String): Unit = {
-    val configConciseWithoutJson = config.root().render(ConfigRenderOptions.concise().setFormatted(true).setJson(false))
-    writeStringToFile(configConciseWithoutJson, new File(outputDirectory, "fullBeamConfig.conf"))
 
-    writeStringToFile(config.root().render(), new File(outputDirectory, "fullBeamConfigJson.conf"))
+    val configResolveOptions = ConfigResolveOptions.defaults().setAllowUnresolved(true)
+
+    val templateConf = ConfigFactory
+      .parseResources("beam-template.conf")
+      .withoutPath("matsim.modules.vehicles.vehiclesFile")
+      .withoutPath("matsim.modules.transit.vehiclesFile")
+      .withoutPath("matsim.modules.counts.inputCountsFile")
+      .withoutPath("matsim.modules.strategy.planSelectorForRemoval")
+      .resolve(configResolveOptions)
+    val fullConfig = config.resolve().withFallback(templateConf).resolve()
+
+    val defaultConfig = fullConfig
+      .entrySet()
+      .asScala
+      .collect {
+        case entry if addKey(entry.getValue.unwrapped) =>
+          val unwrapped = entry.getValue.unwrapped()
+          val paramValue = unwrapped.toString
+          if (paramValue.contains("|")) {
+            entry.getKey -> actualValue(paramValue)
+          } else {
+            entry.getKey -> unwrapped
+          }
+      }
+      .toMap
+      .asJava
+    val defaultValues = ConfigFactory.parseMap(defaultConfig).resolve()
+    val configConciseWithoutJson =
+      defaultValues.root().render(ConfigRenderOptions.concise().setFormatted(true).setJson(false))
+    writeStringToFile(configConciseWithoutJson, new File(outputDirectory, "fullBeamConfig.conf"))
+    writeStringToFile(defaultValues.root().render(), new File(outputDirectory, "fullBeamConfigJson.conf"))
+  }
+
+  private def actualValue(paramValue: String): Any = {
+    val value = paramValue.substring(paramValue.lastIndexOf('|') + 1).trim
+    if (paramValue.contains("int")) {
+      return value.toInt
+    }
+    if (paramValue.contains("double")) {
+      return value.toDouble
+    }
+    if (paramValue.contains("boolean")) {
+      return value.toBoolean
+    }
+    value
+  }
+
+  private def addKey(value: AnyRef): Boolean = {
+    if ("int?" == value.toString || "double?" == value.toString || "[double]" == value.toString) {
+      return false
+    }
+    true
   }
 
   private def writeStringToFile(text: String, output: File): Unit = {

--- a/src/main/scala/beam/sim/BeamHelper.scala
+++ b/src/main/scala/beam/sim/BeamHelper.scala
@@ -1,7 +1,6 @@
 package beam.sim
 
 import java.io.FileOutputStream
-import java.io.{File, FileOutputStream, FileWriter, PrintWriter}
 import java.nio.file.{Files, Paths, StandardCopyOption}
 import java.time.ZonedDateTime
 import java.util.Properties
@@ -39,14 +38,7 @@ import com.conveyal.r5.transit.TransportNetwork
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.module.scala.DefaultScalaModule
 import com.google.inject
-import com.typesafe.config.{
-  ConfigFactory,
-  ConfigRenderOptions,
-  ConfigResolveOptions,
-  ConfigValueFactory,
-  ConfigValueType,
-  Config => TypesafeConfig
-}
+import com.typesafe.config.{ConfigFactory, Config => TypesafeConfig}
 import com.typesafe.scalalogging.LazyLogging
 import kamon.Kamon
 import org.matsim.api.core.v01.{Id, Scenario}
@@ -659,86 +651,11 @@ trait BeamHelper extends LazyLogging {
 
     prepareDirectories(config, beamConfig, outputDirectory)
 
-    writeFullConfigs(config, outputDirectory)
+    ConfigHelper.writeFullConfigs(config, outputDirectory)
 
     val matsimConfig: MatsimConfig = buildMatsimConfig(config, beamConfig, outputDirectory)
 
     BeamExecutionConfig(beamConfig, matsimConfig, outputDirectory)
-  }
-
-  /**
-    * This method merges all configuration parameters into a single file including parameters from
-    * 'include' statements. Two full config files are written out: One without comments and one with
-    * comments in JSON format.
-    * @param config the input config file
-    * @param outputDirectory output folder where full configs will be generated
-    */
-  private def writeFullConfigs(config: TypesafeConfig, outputDirectory: String): Unit = {
-
-    val configResolveOptions = ConfigResolveOptions.defaults().setAllowUnresolved(true)
-
-    val templateConf = ConfigFactory
-      .parseResources("beam-template.conf")
-      .withoutPath("matsim.modules.vehicles.vehiclesFile")
-      .withoutPath("matsim.modules.transit.vehiclesFile")
-      .withoutPath("matsim.modules.counts.inputCountsFile")
-      .withoutPath("matsim.modules.strategy.planSelectorForRemoval")
-      .resolve(configResolveOptions)
-    val fullConfig = config.resolve().withFallback(templateConf).resolve()
-
-    val defaultConfig = fullConfig
-      .entrySet()
-      .asScala
-      .collect {
-        case entry if shouldAddKey(entry.getValue.unwrapped) =>
-          val unwrapped = entry.getValue.unwrapped()
-          val paramValue = unwrapped.toString
-          if (paramValue.contains("|")) {
-            entry.getKey -> actualValue(paramValue)
-          } else {
-            entry.getKey -> unwrapped
-          }
-      }
-      .toMap
-      .asJava
-    val defaultValues = ConfigFactory.parseMap(defaultConfig).resolve()
-    val configConciseWithoutJson =
-      defaultValues.root().render(ConfigRenderOptions.concise().setFormatted(true).setJson(false))
-    writeStringToFile(configConciseWithoutJson, new File(outputDirectory, "fullBeamConfig.conf"))
-    writeStringToFile(defaultValues.root().render(), new File(outputDirectory, "fullBeamConfigJson.conf"))
-  }
-
-  private def actualValue(paramValue: String): Any = {
-    val value = paramValue.substring(paramValue.lastIndexOf('|') + 1).trim
-    if (paramValue.contains("int")) {
-      return value.toInt
-    }
-    if (paramValue.contains("double")) {
-      if ("Double.PositiveInfinity" == value) {
-        return Double.PositiveInfinity
-      }
-      if ("Double.NegativeInfinity" == value) {
-        return Double.NegativeInfinity
-      }
-      return value.toDouble
-    }
-    if (paramValue.contains("boolean")) {
-      return value.toBoolean
-    }
-    value
-  }
-
-  private def shouldAddKey(value: AnyRef): Boolean = {
-    if ("int?" == value.toString || "double?" == value.toString || "[double]" == value.toString) {
-      return false
-    }
-    true
-  }
-
-  private def writeStringToFile(text: String, output: File): Unit = {
-    val fileWriter = new PrintWriter(output)
-    fileWriter.write(text)
-    fileWriter.close
   }
 
   protected def buildNetworkCoordinator(beamConfig: BeamConfig): NetworkCoordinator = {

--- a/src/main/scala/beam/sim/config/BeamConfig.scala
+++ b/src/main/scala/beam/sim/config/BeamConfig.scala
@@ -890,9 +890,7 @@ object BeamConfig {
                 c: com.typesafe.config.Config
               ): BeamConfig.Beam.Agentsim.Agents.RideHail.Initialization.Parking = {
                 BeamConfig.Beam.Agentsim.Agents.RideHail.Initialization.Parking(
-                  filePath =
-                    if (c.hasPathOrNull("filePath")) c.getString("filePath")
-                    else "/test/input/beamville/ride-hail-parking.csv"
+                  filePath = if (c.hasPathOrNull("filePath")) c.getString("filePath") else ""
                 )
               }
             }

--- a/src/main/scala/beam/sim/config/BeamConfig.scala
+++ b/src/main/scala/beam/sim/config/BeamConfig.scala
@@ -957,9 +957,7 @@ object BeamConfig {
 
             def apply(c: com.typesafe.config.Config): BeamConfig.Beam.Agentsim.Agents.RideHail.Initialization = {
               BeamConfig.Beam.Agentsim.Agents.RideHail.Initialization(
-                filePath =
-                  if (c.hasPathOrNull("filePath")) c.getString("filePath")
-                  else "/test/input/beamville/ride-hail-fleet.csv",
+                filePath = if (c.hasPathOrNull("filePath")) c.getString("filePath") else "",
                 initType = if (c.hasPathOrNull("initType")) c.getString("initType") else "PROCEDURAL",
                 parking = BeamConfig.Beam.Agentsim.Agents.RideHail.Initialization.Parking(
                   if (c.hasPathOrNull("parking")) c.getConfig("parking")
@@ -1515,9 +1513,7 @@ object BeamConfig {
               if (c.hasPathOrNull("filePath")) c.getString("filePath") else "/test/input/beamville/taz-centers.csv",
             parkingCostScalingFactor =
               if (c.hasPathOrNull("parkingCostScalingFactor")) c.getDouble("parkingCostScalingFactor") else 1.0,
-            parkingFilePath =
-              if (c.hasPathOrNull("parkingFilePath")) c.getString("parkingFilePath")
-              else "/test/input/beamville/taz-parking.csv",
+            parkingFilePath = if (c.hasPathOrNull("parkingFilePath")) c.getString("parkingFilePath") else "",
             parkingManager = BeamConfig.Beam.Agentsim.Taz.ParkingManager(
               if (c.hasPathOrNull("parkingManager")) c.getConfig("parkingManager")
               else com.typesafe.config.ConfigFactory.parseString("parkingManager{}")
@@ -1641,9 +1637,7 @@ object BeamConfig {
             averageCountsOverIterations =
               if (c.hasPathOrNull("averageCountsOverIterations")) c.getInt("averageCountsOverIterations") else 1,
             countsScaleFactor = if (c.hasPathOrNull("countsScaleFactor")) c.getInt("countsScaleFactor") else 10,
-            inputCountsFile =
-              if (c.hasPathOrNull("inputCountsFile")) c.getString("inputCountsFile")
-              else "/test/input/beamville/counts.xml",
+            inputCountsFile = if (c.hasPathOrNull("inputCountsFile")) c.getString("inputCountsFile") else "",
             writeCountsInterval = if (c.hasPathOrNull("writeCountsInterval")) c.getInt("writeCountsInterval") else 1
           )
         }

--- a/src/main/scala/beam/sim/config/ConfigHelper.scala
+++ b/src/main/scala/beam/sim/config/ConfigHelper.scala
@@ -1,0 +1,87 @@
+package beam.sim.config
+
+import java.io.{File, PrintWriter}
+
+import com.typesafe.config.{ConfigRenderOptions, ConfigResolveOptions, ConfigFactory, Config => TypesafeConfig}
+import scala.collection.JavaConverters._
+
+object ConfigHelper {
+
+  private val skipConfigParams = List(
+    "matsim.modules.vehicles.vehiclesFile",
+    "matsim.modules.transit.vehiclesFile",
+    "matsim.modules.counts.inputCountsFile",
+    "matsim.modules.strategy.planSelectorForRemoval"
+  )
+
+  /**
+    * This method merges all configuration parameters into a single file including parameters from
+    * 'include' statements. Two full config files are written out: One without comments and one with
+    * comments in JSON format.
+    * @param config the input config file
+    * @param outputDirectory output folder where full configs will be generated
+    */
+  def writeFullConfigs(config: TypesafeConfig, outputDirectory: String): Unit = {
+
+    val configResolveOptions = ConfigResolveOptions.defaults().setAllowUnresolved(true)
+    var templateConf = ConfigFactory.parseResources("beam-template.conf")
+    skipConfigParams.foreach(configKey => {
+      templateConf = templateConf.withoutPath(configKey)
+    })
+    val fullConfig = config.resolve().withFallback(templateConf.resolve(configResolveOptions)).resolve()
+
+    val defaultConfig = fullConfig
+      .entrySet()
+      .asScala
+      .collect {
+        case entry if shouldAddKey(entry.getValue.unwrapped) =>
+          val unwrapped = entry.getValue.unwrapped()
+          val paramValue = unwrapped.toString
+          if (paramValue.contains("|")) {
+            entry.getKey -> actualValue(paramValue)
+          } else {
+            entry.getKey -> unwrapped
+          }
+      }
+      .toMap
+      .asJava
+    val defaultValues = ConfigFactory.parseMap(defaultConfig).resolve()
+    val configConciseWithoutJson =
+      defaultValues.root().render(ConfigRenderOptions.concise().setFormatted(true).setJson(false))
+    writeStringToFile(configConciseWithoutJson, new File(outputDirectory, "fullBeamConfig.conf"))
+    writeStringToFile(defaultValues.root().render(), new File(outputDirectory, "fullBeamConfigJson.conf"))
+  }
+
+  private def actualValue(paramValue: String): Any = {
+    val value = paramValue.substring(paramValue.lastIndexOf('|') + 1).trim
+    if (paramValue.contains("int")) {
+      return value.toInt
+    }
+    if (paramValue.contains("double")) {
+      if ("Double.PositiveInfinity" == value) {
+        return Double.PositiveInfinity
+      }
+      if ("Double.NegativeInfinity" == value) {
+        return Double.NegativeInfinity
+      }
+      return value.toDouble
+    }
+    if (paramValue.contains("boolean")) {
+      return value.toBoolean
+    }
+    value
+  }
+
+  private def shouldAddKey(value: AnyRef): Boolean = {
+    if ("int?" == value.toString || "double?" == value.toString || "[double]" == value.toString) {
+      return false
+    }
+    true
+  }
+
+  private def writeStringToFile(text: String, output: File): Unit = {
+    val fileWriter = new PrintWriter(output)
+    fileWriter.write(text)
+    fileWriter.close
+  }
+}

